### PR TITLE
FCFunction correct the <random> and <urandom> whit the <random> c++ c…

### PR DESCRIPTION
**Pull request by abassign**
…lass

### CODE REVIEW

File: src/math/FGFunction.cpp [hunk 1]
```diff
@@ -29,6 +29,8 @@ INCLUDES
 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%*/
 
 #include <iomanip>
+#include <random>
+#include <chrono>
```
Comment (bcoconni) [inline src/math/FGFunction.cpp]: You must include `<memory>` for the compilation to succeed on Linux.
```suggestion
#include <memory>

```

**Comment by bcoconni** [comment]
> These changes don't allow the user to specify a seed value.

Right. This can be worked around with an attribute to the `<random>` function:
```xml
<random seed="123456"/>
```
In order to match that requirement, the C++ code must be modified:
```diff
    } else if (operation == "random") {
+      string seed_attr = element->GetAttributeValue("seed");
+      unsigned int seed;
+      if (seed_attr.empty())
+        seed = chrono::system_clock::now().time_since_epoch().count();
+      else
+        seed = atof(seed_attr.c_str());
-      unsigned int seed = chrono::system_clock::now().time_since_epoch().count();
      shared_ptr<default_random_engine> generator(new default_random_engine(seed));
      shared_ptr<normal_distribution<double>> distribution(new normal_distribution<double>(0.0, 1.0));
      auto f = [generator, distribution](const decltype(Parameters)& p)->double {
                 return (*distribution.get())(*generator);
               };
      Parameters.push_back(new aFunc<decltype(f), 0>(f, fdmex, element, Prefix, var));
```

**Comment by seanmcleod** [comment]
`atoi` rather than `atof` right?

> +        seed = atof(seed_attr.c_str());

**Comment by bcoconni** [comment]
> `atoi` rather than `atof` right?

Oops ! You are right it should be `atoi`.

In addition the functions must respond to the property `simulation/randomseed` so the code could be something like
```diff
    } else if (operation == "random") {
-      unsigned int seed = chrono::system_clock::now().time_since_epoch().count();
+      string seed_attr = element->GetAttributeValue("seed");
+      unsigned int seed;
+      if (seed_attr.empty())
+        seed = fdmex->SRand(); // Use the seed from simulation/randomseed by default
+      else if (seed_attr == "time_now")
+        seed = chrono::system_clock::now().time_since_epoch().count();
+      else
+        seed = atoi(seed_attr.c_str());
      shared_ptr<default_random_engine> generator(new default_random_engine(seed));
      shared_ptr<normal_distribution<double>> distribution(new normal_distribution<double>(0.0, 1.0));
      auto f = [generator, distribution](const decltype(Parameters)& p)->double {
                 return (*distribution.get())(*generator);
               };
      Parameters.push_back(new aFunc<decltype(f), 0>(f, fdmex, element, Prefix, var));
```
But then you have to make the method `FGFDMExec::SRand()` public in `src/FGFDMExec.h`
https://github.com/JSBSim-Team/jsbsim/blob/fa1f7cd3f92b8038f916d5139b8c5dcdf9b9864c/src/FGFDMExec.h#L667

With the code above, there are 3 options for the attribute :
1. No `seed` attribute is provided then the random function will use the "global" seed from the property `simulation/seed`
```xml
<random/>
```
2. The `seed` attribute is provided with the value `time_now` then the random function will use a seed based on the system clock.
```xml
<random seed="time_now"/>
```
3. The `seed` attribute is provided with a number then the random function will use the provided number as a seed.
```xml
<random seed="123456"/>
```

**Comment by bcoconni** [comment]
Also for the record, this feature request is based on the discussion from the issue #232.

**Comment by abassign** [comment]
It's a very interesting change, I like it, but I'd also like to be able to enter the two additional parameters that define the interval. For my simulations I often have to intervene with a multiplication to have the values in the range.

**Comment by bcoconni** [comment]
> I'd also like to be able to enter the two additional parameters that define the interval.

No problem, you can apply the same principle for new attributes `mean` and `stddev`:
```c++
      double mean = 0.0;
      double stddev = 1.0;
      string mean_attr = element->GetAttributeValue("mean");
      string stddev_attr = element->GetAttributeValue("stddev");
      if (!mean_attr.empty())
        mean = atof(mean_attr.c_str());
      if (!stddev_attr.empty())
        stddev = atof(stddev_attr.c_str());
```
The function could then be specified as
```xml
<random seed="time_now" mean="-1.0" stddev="1.5"/>
```

**Comment by abassign** [comment]
@bcoconni I did the pull-request, so the change to the additional parameters who can do it?

**Comment by bcoconni** [comment]
Just push new commits to your branch `abassign:jsb-abassign` and GitHub will take care automatically of the rest and update this pull request.

File: src/math/FGFunction.cpp (hunk 1 above)
Comment (abassign) [reply src/math/FGFunction.cpp]: I put the #include <memory> i my last version

File: src/FGFDMExec.h [hunk 2]
```diff
@@ -84,7 +84,7 @@ CLASS DOCUMENTATION
     executive is subsequently directed to load the chosen aircraft specification
     file:
 
-    @code{.cpp}
+    @code
```
Comment (bcoconni) [inline src/FGFDMExec.h]: Please don't remove the `{.cpp}` elements
```suggestion
    @code{.cpp}
```

File: src/FGFDMExec.h [hunk 3]
```diff
@@ -100,15 +100,15 @@ CLASS DOCUMENTATION
     from JSBSim. The state variables are used to drive the instrument displays
     and to place the vehicle model in world space for visual rendering:
 
-    @code{.cpp}
+    @code
```
Comment (bcoconni) [inline src/FGFDMExec.h]: Please don't remove the `{.cpp}` elements
```suggestion
    @code{.cpp}
```

File: src/FGFDMExec.h [hunk 4]
```diff
@@ -100,15 +100,15 @@ CLASS DOCUMENTATION
...
     copy_to_JSBsim(); // copy control inputs to JSBSim
     fdmex->RunIC(); // loop JSBSim once w/o integrating
     copy_from_JSBsim(); // update the bus
     @endcode
 
     Once initialization is complete, cyclic execution proceeds:
 
-    @code{.cpp}
+    @code
```
Comment (bcoconni) [inline src/FGFDMExec.h]: Please don't remove the `{.cpp}` elements
```suggestion
    @code{.cpp}
```

File: src/FGFDMExec.h [hunk 5]
```diff
@@ -123,7 +123,7 @@ CLASS DOCUMENTATION
     file can be supplied to the stub program. Scripting (see FGScript) provides
     a way to supply command inputs to the simulation:
 
-    @code{.cpp}
+    @code
```
Comment (bcoconni) [inline src/FGFDMExec.h]: Please don't remove the `{.cpp}` elements
```suggestion
    @code{.cpp}
```

File: src/math/FGFunction.h [hunk 6]
```diff
@@ -689,10 +688,24 @@ refers to one or more instances of a property, value, or table.
      </switch>
      @endcode
 - @b random Takes no arguments and returns a Gaussian distributed random number
-    @code <random/> @endcode
+    @code
+    <random/> 
+    Default configuration with mean: mean="0.0" and standard deviation: stddev="1.0"
```
Comment (bcoconni) [inline src/math/FGFunction.h]: Do not include comments inside `@code` / `@endcode` pairs
```suggestion
    @endcode
    Default configuration with mean: mean="0.0" and standard deviation: stddev="1.0"
```

File: src/math/FGFunction.h [hunk 7]
```diff
@@ -689,10 +688,24 @@ refers to one or more instances of a property, value, or table.
...
+    <random seed="1234"/>
+    <random seed="time_now"/>
+    <random seed="time_now" mean="0.0" stddev="1.0"/>
+    @endcode
 - @b urandom, Takes no arguments and returns a uniformly distributed random
               number between -1 and +1
-    @code<urandom/>@endcode
+    @code
+    <urandom/>
+    Default configuration with lower value: lower="-1.0" and upper value: upper="1.0"
```
Comment (bcoconni) [inline src/math/FGFunction.h]: Do not include comments inside `@code` / `@endcode` pairs
```suggestion
    @endcode
    Default configuration with lower value: lower="-1.0" and upper value: upper="1.0"
```

File: src/math/FGFunction.h [hunk 8]
```diff
@@ -689,10 +688,24 @@ refers to one or more instances of a property, value, or table.
...
+    Optional parameters:
+    <random seed="1234"/>
```
Comment (bcoconni) [inline src/math/FGFunction.h]: Do not include comments inside `@code` / `@endcode` pairs
```suggestion
    @code
    <random seed="1234"/>
```

File: src/FGFDMExec.h (hunk 2 above)
Comment (abassign) [reply src/FGFDMExec.h]: Ops ... Sorry, in fact I noticed it after I had committed the correction and I immediately correct it.

**Comment by abassign** [comment]
Beautifull :) I hope that as soon as possible it will also be in FGFS so I can publish the new version of G91R1B with <random>.
Best regards
Adriano

**Pull request merged**